pgxn/neon/pagestore_smgr.c

```diff
@@ -4425,13 +4529,20 @@ neon_redo_read_buffer_filter(XLogReaderState *record, uint8 block_id)
 	}
 
 	/*
-	 * we don't have the buffer in memory, update lwLsn past this record, also
-	 * evict page from file cache
+	 * If a buffer is present in LFC, we should probably replay from there
+	 * (rather than evict it).
+	 */
+	if (no_redo_needed)
+	{
+		no_redo_needed = !lfc_cache_contains_prewarm(rinfo, forknum, blkno);
+	}
+
+	/*
+	 * we don't have the buffer in memory, so update lwLsn past this record
 	 */
 	if (no_redo_needed)
 	{
 		SetLastWrittenLSNForBlock(end_recptr, rinfo, forknum, blkno);
-		lfc_evict(rinfo, forknum, blkno);
```

Comment:
I wonder why we do not need to perform eviction from LFC in this case? There is no check for LwLSN in `neon_read` so we can actually read deteriorated value.

Reply:
Applying changes happen not only during prewarm. We can have normal hot-standby replica, executing read-only queries and performing prefetch. Ok, it is mostly matter of naming - `lfc_cache_contains_prewarm` seems to be not correct in general case.

But there seems to be no locking between `lfc_cache_contains_prewarm` and `SetLastWrittenLSNForBlock`. What will happen if `lfc_cache_contains_prewarm`  returns false because prewarming of the chunk is not yet started, but then it is started and load some page. Then we skip applying WAL to this page and as a result have deteriorated version of the page in LFC.

Reply:
I think that we should first (unconditionally) update LwLSN and only after it  check presence of the page in LFC cache. After update of LwLSN, any attempt to prefetch page with smaller LSN will be rejected.